### PR TITLE
Scheduled maintenance message Fixes:#434

### LINKED ISSUE #434: Future Maintenance Notification Message
Add a way to notify maintenance schedule on GraphSpace using django/bash commands with environment variables.

### PR DESCRIPTION
**Pull request by adbharadwaj**
## Purpose
We want to be able to update the server to display a warning message to users when we plan to do heavy maintenance that could affect the user's work. If a graph is open and being currently edited by users when maintenance is live, they may lose those unsaved changes.

Example:
Fixes #434  .

## Approach
Uses a script to change environment variables for a Boolean variable for existence of scheduled maintenance, and 2 variables for a start and end date of the maintenance. The environment variables are loaded into the project via base.py and a partial was created to display the yellow warning message on the top of the page.

## Learning
image

#### Blog Posts
- [How to Pull Request](https://github.com/flexyford/pull-request) Github Repo with Learning focused Pull Request Template.

### CODE REVIEW

File: scripts/schedule_maintenance.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+#run with "source schedule_maintenance.sh [arg1] [arg2] [arg3]"
+#pass dates as "m/d/y@hour:min"
+if (( $# > 0 ))
+then
+  if (($[$1 == true] || $[$1 == false]))
+  then
+    export IS_MAINTENANCE_SCHEDULED=$1
+    echo $IS_MAINTENANCE_SCHEDULED
```
Comment (adbharadwaj) [inline scripts/schedule_maintenance.sh:9]: Maybe change this to a proper log message like `Scheduled Maintenance Message has been enabled/disabled`?

File: scripts/schedule_maintenance.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    printf "The value of arg1:IS_MAINTENANCE_SCHEDULED must be a boolean\n"
+  fi
+else
+  printf "Not enough arguements. Please enter arguements [IS_MAINTENANCE_SCHEDULED] [MAINTENANCE_START_DATETIME] [MAINTENANCE_END_DATETIME]\n"
+fi
+
+if (( $# > 1))
+then
+  export MAINTENANCE_START_DATETIME=$2
+  echo $MAINTENANCE_START_DATETIME
```
Comment (adbharadwaj) [inline scripts/schedule_maintenance.sh:20]: Similar: Maybe change this to a proper log message like `Scheduled Maintenance starts at $MAINTENANCE_START_DATETIME`

File: scripts/schedule_maintenance.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+fi
+
+if (( $# > 2))
+then
+  export MAINTENANCE_END_DATETIME=$3
+  echo $MAINTENANCE_END_DATETIME
```
Comment (adbharadwaj) [inline scripts/schedule_maintenance.sh:26]: Similar: Maybe change this to a proper log message

**Review by adbharadwaj — COMMENTED** [review]
Can you add some description for the PR. Use the default PR description template (Purpose, Approach, etc).

Did you confirm that the scheduled maintenance message appears on all pages if enabled?

File: templates/partials/scheduled_maintenance.html [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<!DOCTYPE html>
+
+{% load staticfiles %}
+
+{% if is_maintenance_scheduled and maintenance_start_datetime and maintenance_end_datetime%}
+<div class="alert alert-warning" role="alert">
+  There will be future maintenance on {{ maintenance_start_datetime }} until {{ maintenance_end_datetime }}
```
Comment (adbharadwaj) [inline templates/partials/scheduled_maintenance.html:7]: You should change this to something like 

> Please note, there will be scheduled downtime on {{ maintenance_start_datetime }} until {{ maintenance_end_datetime }}. We will be using this time to update our infrastructure.
> 
> **Any graph updated or uploaded during the maintenance period will be lost. Please plan your work accordingly.**
> 
> We appreciate your patience and apologize for any inconvenience.

File: scripts/schedule_maintenance.sh (hunk 1 above)
Comment (jddanna) [reply scripts/schedule_maintenance.sh:9]: okay. I'll change the script responses.

File: templates/partials/scheduled_maintenance.html (hunk 4 above)
Comment (jddanna) [reply templates/partials/scheduled_maintenance.html:7]: I changed the warning message to look like this:
image

File: scripts/schedule_maintenance.sh (hunk 1 above)
Comment (jddanna) [reply scripts/schedule_maintenance.sh]: and yes the warning displays on home page, about us page, and graph page.

**Comment by jddanna** [comment]
> Please make some minor changes: "Please note that GraphSpace will be unavailable due to scheduled maintenance from {{ maintenance_start_datetime }} to {{ maintenance_end_datetime }}."

Here's an updated view
image

**Review by jddanna — APPROVED** [review]

**Pull request merged**